File: harness/authenticated_adr_approval_contract/proposal.py
"""Exact strict-Proposed ADR v2 byte and digest binding."""

from __future__ import annotations

import base64
import hashlib
import re
from typing import Any

from architecture_decision_record_v2 import validate_document_bytes
from governance_contract import ContractError as ADRV2Error

from .canonical import ContractError, bounded_canonical_json, self_digest
from .constants import (CANONICALIZATION, MAX_PROPOSAL_BINDING_BYTES,
                        MAX_PROPOSAL_BYTES, PROFILE_ID, PROPOSAL_BINDING_API,
                        PROPOSAL_BINDING_DOMAIN)
from .shape import decode_base64url, require_keys, sha256

PROPOSAL_BINDING_FIELDS = {
    "adr_id", "api_version", "body_sha256", "canonicalization", "document_name",
    "kind", "physical_sha256", "profile_id", "proposal_binding_sha256",
    "self_sha256", "status",
}


def proposal_binding_sha256(value: dict[str, Any]) -> str:
    return self_digest(PROPOSAL_BINDING_DOMAIN, value, ("proposal_binding_sha256",),
                       MAX_PROPOSAL_BINDING_BYTES, "ArchitectureDecisionProposalBinding")


def validate_proposal_binding(value: Any) -> dict[str, Any]:
    label = "ArchitectureDecisionProposalBinding"
    node = require_keys(value, label, PROPOSAL_BINDING_FIELDS)
    bounded_canonical_json(node, MAX_PROPOSAL_BINDING_BYTES, label)
    expected = (PROPOSAL_BINDING_API, CANONICALIZATION, label, PROFILE_ID, "proposed")
    actual = (node["api_version"], node["canonicalization"], node["kind"],
              node["profile_id"], node["status"])
    if actual != expected:
        raise ContractError(f"{label} envelope drifted from v1")
    adr_match = (re.fullmatch(r"ADR-((?!0000)[0-9]{4})", node["adr_id"])
                 if isinstance(node["adr_id"], str) else None)
    if adr_match is None:
        raise ContractError("proposal_binding.adr_id is malformed")
    name_match = (re.fullmatch(r"ADR-([0-9]{4})-[a-z0-9]+(?:-[a-z0-9]+)*\.md",
                              node["document_name"])
                  if isinstance(node["document_name"], str) else None)
    if name_match is None or name_match.group(1) != adr_match.group(1):
        raise ContractError("proposal_binding.document_name is malformed")
    for field in ("body_sha256", "physical_sha256", "proposal_binding_sha256",
                  "self_sha256"):
        sha256(node[field], f"proposal_binding.{field}")
    if node["proposal_binding_sha256"] != proposal_binding_sha256(node):
        raise ContractError("proposal binding self digest does not match")
    return node


def derive_proposal_binding(proposal_bytes: bytes, document_name: str) -> dict[str, Any]:
    if not isinstance(proposal_bytes, bytes) or not 1 <= len(proposal_bytes) <= MAX_PROPOSAL_BYTES:
        raise ContractError("proposal bytes must be non-empty and at most 262144 bytes")
    try:
        metadata = validate_document_bytes(proposal_bytes, document_name)
    except ADRV2Error as error:
        raise ContractError(f"proposal is not a strict Proposed ADR v2: {error}") from error
    binding = {
        "adr_id": metadata["adr_id"],
        "api_version": PROPOSAL_BINDING_API,
        "body_sha256": metadata["body_sha256"],
        "canonicalization": CANONICALIZATION,
        "document_name": metadata["document_name"],
        "kind": "ArchitectureDecisionProposalBinding",
        "physical_sha256": hashlib.sha256(proposal_bytes).hexdigest(),
        "profile_id": PROFILE_ID,
        "proposal_binding_sha256": "",
        "self_sha256": metadata["self_sha256"],
        "status": "proposed",
    }
    binding["proposal_binding_sha256"] = proposal_binding_sha256(binding)
    return validate_proposal_binding(binding)


def validate_proposal_bytes(proposal_bytes: bytes, binding: dict[str, Any]) -> dict[str, Any]:
    validate_proposal_binding(binding)
    derived = derive_proposal_binding(proposal_bytes, binding["document_name"])
    if derived != binding:
        raise ContractError("exact proposal bytes do not match ProposalBinding")
    return derived


def encode_proposal_document(proposal_bytes: bytes) -> str:
    if not isinstance(proposal_bytes, bytes) or not 1 <= len(proposal_bytes) <= MAX_PROPOSAL_BYTES:
        raise ContractError("proposal bytes must be non-empty and at most 262144 bytes")
    return base64.urlsafe_b64encode(proposal_bytes).decode("ascii").rstrip("=")


def decode_proposal_document(value: Any, binding: dict[str, Any],
                             label: str) -> tuple[bytes, dict[str, Any]]:
    proposal_bytes = decode_base64url(value, label, MAX_PROPOSAL_BYTES)
    validate_proposal_bytes(proposal_bytes, binding)
    try:
        metadata = validate_document_bytes(proposal_bytes, binding["document_name"])
    except ADRV2Error as error:
        raise ContractError(f"proposal is not a strict Proposed ADR v2: {error}") from error
    return proposal_bytes, metadata

File: forge-core/internal/authenticatedadrapprovalcontract/proposal.go
package authenticatedadrapprovalcontract

import (
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"fmt"
	"regexp"

	"forgeos/forge-core/internal/adrv2"
)

var proposalNamePattern = regexp.MustCompile(`^ADR-([0-9]{4})-[a-z0-9]+(?:-[a-z0-9]+)*\.md$`)

func validateProposalBinding(value any) (map[string]any, error) {
	label := "ArchitectureDecisionProposalBinding"
	fields := []string{"adr_id", "api_version", "body_sha256", "canonicalization",
		"document_name", "kind", "physical_sha256", "profile_id",
		"proposal_binding_sha256", "self_sha256", "status"}
	node, err := requireKeys(value, label, fields...)
	if err != nil {
		return nil, err
	}
	if _, err = boundedCanonicalJSON(node, maxProposalBindingBytes, label); err != nil {
		return nil, err
	}
	if node["api_version"] != proposalBindingAPI || node["canonicalization"] != canonicalization ||
		node["kind"] != label || node["profile_id"] != profileID || node["status"] != "proposed" {
		return nil, fmt.Errorf("%s envelope drifted from v1", label)
	}
	if err = validateProposalBindingIdentity(node); err != nil {
		return nil, err
	}
	for _, field := range []string{"body_sha256", "physical_sha256", "proposal_binding_sha256", "self_sha256"} {
		if _, err = shaValue(node[field], "proposal_binding."+field); err != nil {
			return nil, err
		}
	}
	digest, err := proposalBindingSHA256(node)
	if err != nil || node["proposal_binding_sha256"] != digest {
		return nil, fmt.Errorf("proposal binding self digest does not match")
	}
	return node, nil
}

func validateProposalBindingIdentity(node map[string]any) error {
	adrID, ok := node["adr_id"].(string)
	if !ok || !validProposalADRID(adrID) {
		return fmt.Errorf("proposal_binding.adr_id is malformed")
	}
	name, ok := node["document_name"].(string)
	match := proposalNamePattern.FindStringSubmatch(name)
	if !ok || match == nil || match[1] != adrID[4:] {
		return fmt.Errorf("proposal_binding.document_name is malformed")
	}
	return nil
}

func validProposalADRID(value string) bool {
	if len(value) != 8 || value[:4] != "ADR-" || value == "ADR-0000" {
		return false
	}
	for _, character := range value[4:] {
		if character < '0' || character > '9' {
			return false
		}
	}
	return true
}

func deriveProposalBinding(proposal []byte, documentName string) (map[string]any, proposalMetadata, error) {
	metadata, err := validateProposalDocument(proposal, documentName)
	if err != nil {
		return nil, proposalMetadata{}, err
	}
	physical := sha256.Sum256(proposal)
	binding := map[string]any{
		"adr_id": metadata.ADRID, "api_version": proposalBindingAPI,
		"body_sha256": metadata.BodySHA256, "canonicalization": canonicalization,
		"document_name": metadata.DocumentName, "kind": "ArchitectureDecisionProposalBinding",
		"physical_sha256": hex.EncodeToString(physical[:]), "profile_id": profileID,
		"proposal_binding_sha256": "", "self_sha256": metadata.SelfSHA256,
		"status": "proposed",
	}
	digest, err := proposalBindingSHA256(binding)
	if err != nil {
		return nil, proposalMetadata{}, err
	}
	binding["proposal_binding_sha256"] = digest
	_, err = validateProposalBinding(binding)
	return binding, metadata, err
}

func validateProposalDocument(proposal []byte, documentName string) (proposalMetadata, error) {
	if len(proposal) == 0 || len(proposal) > maxProposalBytes {
		return proposalMetadata{}, fmt.Errorf("proposal bytes must contain 1..%d bytes", maxProposalBytes)
	}
	document, err := adrv2.ValidateDocument(documentName, proposal)
	if err != nil {
		return proposalMetadata{}, fmt.Errorf("proposal is not a strict Proposed ADR v2: %w", err)
	}
	frontmatter := document.Frontmatter
	return proposalMetadata{ADRID: frontmatter.ADRID,
		ApproverRefs: append([]string(nil), frontmatter.ApproverRefs...),
		BodySHA256:   frontmatter.BodySHA256, DocumentName: frontmatter.DocumentName,
		ExpiresAtUnixMS:  copyInt64Pointer(frontmatter.ExpiresAtUnixMS),
		OwnerRefs:        append([]string(nil), frontmatter.OwnerRefs...),
		ProposedAtUnixMS: frontmatter.ProposedAtUnixMS, SelfSHA256: frontmatter.SelfSHA256,
		Status: frontmatter.Status}, nil
}

func validateProposalBytes(proposal []byte, binding map[string]any) (proposalMetadata, error) {
	if _, err := validateProposalBinding(binding); err != nil {
		return proposalMetadata{}, err
	}
	derived, metadata, err := deriveProposalBinding(proposal, binding["document_name"].(string))
	if err != nil {
		return proposalMetadata{}, err
	}
	if !canonicalEqual(derived, binding) {
		return proposalMetadata{}, fmt.Errorf("exact proposal bytes do not match ProposalBinding")
	}
	return metadata, nil
}

func decodeProposalDocument(value any, binding map[string]any, label string) ([]byte, proposalMetadata, error) {
	proposal, err := decodeBase64URL(value, label, maxProposalBytes)
	if err != nil {
		return nil, proposalMetadata{}, err
	}
	metadata, err := validateProposalBytes(proposal, binding)
	return proposal, metadata, err
}

func encodeProposalDocument(proposal []byte) (string, error) {
	if len(proposal) == 0 || len(proposal) > maxProposalBytes {
		return "", fmt.Errorf("proposal bytes must contain 1..%d bytes", maxProposalBytes)
	}
	return base64.RawURLEncoding.EncodeToString(proposal), nil
}

func copyInt64Pointer(value *int64) *int64 {
	if value == nil {
		return nil
	}
	result := *value
	return &result
}
